File: backend/models/analytics.py
from pydantic import BaseModel


class REMatrixEntry(BaseModel):
    base_out_state: int
    outs: int
    runners_on: str
    expected_runs: float
    occurrences: int


class REMatrixData(BaseModel):
    season: int
    entries: list[REMatrixEntry]


class LinearWeightsRow(BaseModel):
    season: int
    source: str
    wBB: float
    wHBP: float
    w1B: float
    w2B: float
    w3B: float
    wHR: float
    lg_woba: float
    woba_scale: float


class LinearWeightsData(BaseModel):
    season: int
    weights: list[LinearWeightsRow]


class ParkFactorRow(BaseModel):
    team: str
    venue: str
    home_games: int
    road_games: int
    home_rpg: float
    road_rpg: float
    park_factor: float


class ParkFactorsData(BaseModel):
    season: int
    factors: list[ParkFactorRow]

File: frontend/src/types/analytics.ts
export interface REMatrixEntry {
  base_out_state: number;
  outs: number;
  runners_on: string;
  expected_runs: number;
  occurrences: number;
}

export interface REMatrixData {
  season: number;
  entries: REMatrixEntry[];
}

export interface LinearWeightsRow {
  season: number;
  source: string;
  wBB: number;
  wHBP: number;
  w1B: number;
  w2B: number;
  w3B: number;
  wHR: number;
  lg_woba: number;
  woba_scale: number;
}

export interface LinearWeightsData {
  season: number;
  weights: LinearWeightsRow[];
}

export interface ParkFactorRow {
  team: string;
  venue: string;
  home_games: number;
  road_games: number;
  home_rpg: number;
  road_rpg: number;
  park_factor: number;
}

export interface ParkFactorsData {
  season: number;
  factors: ParkFactorRow[];
}

export interface RECountEntry {
  base_out_state: number;
  outs: number;
  runners_on: string;
  balls: number;
  strikes: number;
  expected_runs: number;
  occurrences: number;
}

export interface RECountMatrixData {
  season: number;
  entries: RECountEntry[];
}
